File: daily_solutions/search_insert_position_20250226.py
# Problem: Search Insert Position
"""
Problem Statement:

Given a sorted array of distinct integers and a target value, return the index if the target is found.
If not, return the index where it would be if it were inserted in order.

You must write an algorithm with O(log n) runtime complexity.

Example 1:
Input: nums = [1,3,5,6], target = 5
Output: 2

Example 2:
Input: nums = [1,3,5,6], target = 2
Output: 1

Example 3:
Input: nums = [1,3,5,6], target = 7
Output: 4
"""

def searchInsert(nums: list[int], target: int) -> int:
    """
    Finds the index to insert the target value in a sorted array.

    Args:
        nums: A sorted list of distinct integers.
        target: The target integer to search for or insert.

    Returns:
        The index of the target if found, or the index where it would be inserted.
    """

    low = 0
    high = len(nums) - 1

    while low <= high:
        mid = (low + high) // 2  # Calculate the middle index

        if nums[mid] == target:
            return mid  # Target found at the middle index
        elif nums[mid] < target:
            low = mid + 1  # Target is in the right half
        else:
            high = mid - 1  # Target is in the left half

    return low  # Target not found, return the index where it would be inserted

# Time and Space Complexity:
# Time Complexity: O(log n) - Binary search algorithm.
# Space Complexity: O(1) - Constant space used.

# Test cases
if __name__ == '__main__':
    nums1 = [1, 3, 5, 6]
    target1 = 5
    print(f"Input: nums = {nums1}, target = {target1}, Output: {searchInsert(nums1, target1)}")  # Output: 2

    nums2 = [1, 3, 5, 6]
    target2 = 2
    print(f"Input: nums = {nums2}, target = {target2}, Output: {searchInsert(nums2, target2)}")  # Output: 1

    nums3 = [1, 3, 5, 6]
    target3 = 7
    print(f"Input: nums = {nums3}, target = {target3}, Output: {searchInsert(nums3, target3)}")  # Output: 4

    nums4 = [1, 3, 5, 6]
    target4 = 0
    print(f"Input: nums = {nums4}, target = {target4}, Output: {searchInsert(nums4, target4)}") # Output: 0

    nums5 = [1]
    target5 = 0
    print(f"Input: nums = {nums5}, target = {target5}, Output: {searchInsert(nums5, target5)}") # Output: 0

File: daily_solutions/search_insert_position_20250226.java
// Problem: Search Insert Position
java
/**
 * LeetCode Problem: Search Insert Position
 *
 * Problem Statement:
 * Given a sorted array of distinct integers and a target value, return the index if the target is found.
 * If not, return the index where it would be if it were inserted in order.
 *
 * You must write an algorithm with O(log n) runtime complexity.
 */

public class search_insert_position_20240127 {

    /**
     * Searches for the target value in the sorted array and returns its index if found.
     * If the target is not found, returns the index where it would be inserted in order.
     *
     * @param nums   The sorted array of distinct integers.
     * @param target The target value to search for.
     * @return The index of the target if found, or the index where it would be inserted.
     */
    public int searchInsert(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2; // Prevents potential overflow

            if (nums[mid] == target) {
                return mid; // Target found at index mid
            } else if (nums[mid] < target) {
                left = mid + 1; // Target is in the right half
            } else {
                right = mid - 1; // Target is in the left half
            }
        }

        return left; // Target not found, return the index where it would be inserted
    }

    public static void main(String[] args) {
        search_insert_position_20240127 solution = new search_insert_position_20240127();

        // Test cases
        int[] nums1 = {1, 3, 5, 6};
        int target1 = 5;
        System.out.println("Test Case 1: " + solution.searchInsert(nums1, target1)); // Output: 2

        int[] nums2 = {1, 3, 5, 6};
        int target2 = 2;
        System.out.println("Test Case 2: " + solution.searchInsert(nums2, target2)); // Output: 1

        int[] nums3 = {1, 3, 5, 6};
        int target3 = 7;
        System.out.println("Test Case 3: " + solution.searchInsert(nums3, target3)); // Output: 4

        int[] nums4 = {1, 3, 5, 6};
        int target4 = 0;
        System.out.println("Test Case 4: " + solution.searchInsert(nums4, target4)); // Output: 0
    }

    /**
     * Time Complexity: O(log n) - Binary search reduces the search space by half in each iteration.
     * Space Complexity: O(1) - Constant space is used, regardless of the input size.  Only a few variables (left, right, mid) are used.
     */
}
